### PR TITLE
Fetch PFS list from ServiceRegistry

### LINKED ISSUE #253: I want to get a list of PFS services from the service registry
## User Story

As Chris I want to get a list of PFS services from the service registry so that I can use any PFS service in the network
See [Raiden Client implementation](https://github.com/raiden-network/raiden/commit/bce57dad6c8d183387d658065b7c0838544a8bfa)

## Acceptance criteria
- At start-up, the service registry is queried
- The first one is stored in the config
- It is an asynchronous operation
- The whole config is stored in the state (#399)

### PR DESCRIPTION
**Pull request by andrevmatos**
Fix #253 

- `config.pfs` can be 3 values:
  - undefined, currently the default, enables fetching pfs list from service registry
  - null, disable PFS if one isn't specified explicitly in `findRoutes` or `transfer` options
  - a string, which will be used as pfs URL to be used
- on the pfs list fetching, ServiceRegistry is monitored, and one can use `Raiden.findPFS` public API to fetch a list of available PFSs
  - in this case, the selected PFS can be passed to `findRoutes`, and this PFS will be used by findRoutes
  - if not specified, `findRoutes` will fetch, sort & select automatically a PFS from the registry

Notice that, currently, only the paid PFSs are registered on-chain, and we don't support paying PFS yet (#254), so findRoutes will fail when the PFS request fails due to the lack of an IOU. Workaround is to set the free PFS explicity with `raiden.updateConfig({ pfs: 'free-pfs' })`, if not set yet, on console.

### CODE REVIEW

File: raiden-ts/src/path/utils.ts [hunk 1]
```diff
@@ -33,3 +44,105 @@ export function channelCanRoute(
...
+    version: t.string,
+  });
+  // if it's an address, fetch url from ServiceRegistry, else it's already the URL
+  const url$ = Address.is(pfsAddrOrUrl)
+    ? from(serviceRegistryContract.functions.urls(pfsAddrOrUrl))
+    : of(pfsAddrOrUrl);
+  return url$.pipe(
+    withLatestFrom(config$),
+    mergeMap(([url, { httpTimeout }]) => {
+      if (!url || !url.includes('://')) throw new Error(`Invalid URL: ${url}`);
```
Comment (nephix) [inline raiden-ts/src/path/utils.ts]: Do we need to check for `https` as well or can we expect to only receive https URLs from the contract?

File: raiden-ts/src/raiden.ts [hunk 2]
```diff
@@ -41,18 +42,19 @@ import {
 } from 'rxjs/operators';
 
 import './polyfills';
-import { TokenNetworkRegistry } from './contracts/TokenNetworkRegistry';
-import { TokenNetwork } from './contracts/TokenNetwork';
-import { HumanStandardToken } from './contracts/HumanStandardToken';
-
-import TokenNetworkRegistryAbi from './abi/TokenNetworkRegistry.json';
-import TokenNetworkAbi from './abi/TokenNetwork.json';
-import HumanStandardTokenAbi from './abi/HumanStandardToken.json';
```
Comment (nephix) [inline raiden-ts/src/raiden.ts]: Looks like we don't need `src/abi/` anymore? If yes, might make sense to remove the `abi` related code from `scripts/copyContracts.js:preBuild()` as well unless we need it in the future

**Comment by nephix** [comment]
Good job

**Comment by kelsos** [comment]
@nephix My guess would be that if there is no newer RC or release version than the one we were using during testing then it wouldn't make sense to upgrade the contracts.

File: raiden-ts/src/path/utils.ts (hunk 1 above)
Comment (kelsos) [reply raiden-ts/src/path/utils.ts]: @Dominik1999 could you help us here?

Comment (Dominik1999) [reply raiden-ts/src/path/utils.ts]: I'd love to help. Can you elaborate on the question a bit more? Then either me or Karl can answer that, I am sure

File: raiden-ts/src/raiden.ts (hunk 2 above)
Comment (andrevmatos) [reply raiden-ts/src/raiden.ts]: We still need the abis to build the contracts with typechain. We just don't need it at runtime. Unless typechain gained the ability to parse the bundled contracts.json, we either keep it, or remove it after running typechain, but it still needed.

File: raiden-ts/src/path/utils.ts (hunk 1 above)
Comment (nephix) [reply raiden-ts/src/path/utils.ts]: The question is whether PFS servers are always running on HTTPS and the ServiceRegistry always returns HTTPS URLs or if there is any requirement to support HTTP as well

Comment (andrevmatos) [reply raiden-ts/src/path/utils.ts]: We could either require `startsWith('https://')` or add it on our own, seems there's at least one case on the current contracts where only the domain is set (without schema), which is filtered now, but could be useful to be kept instead of filtered out here.

File: raiden-ts/src/path/epics.ts [hunk 3]
```diff
@@ -63,21 +84,46 @@ export const pathFindServiceEpic = (
...
+              } else if (!action.payload.pfs && configPfs === null) {
+                // pfs not specified in action and disabled (null) in config
+                throw new Error(`PFS disabled and no direct route available`);
+              } else {
+                // else, request a route from PFS.
+                // pfs$ - Observable which emits one PFS info and then completes
+                const pfs$ = action.payload.pfs
+                  ? // first, honor action.payload.pfs
+                    of(action.payload.pfs)
+                  : !isNil(configPfs)
```
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden-ts/src/path/epics.ts]: `isNil` actually test if `null` || `undefined`. It stayed because a previous version used it in some more places, including over result of function call, and to do the double check without the helper I'd need to store it in a variable to test, which would be troublesome in the `if`. Now, it's simple enough to be removed, although I think it's already used somewhere else so wouldn't save space anyway.

[the remark this answers is not in the document]
Comment (nephix) [reply raiden-ts/src/path/epics.ts]: `foo == null` also tests `null || undefined` and `isNil` actually uses `==` [under the hood](https://github.com/lodash/lodash/blob/master/isNil.js) as well. Doesn't seem to be used on `lib2` branch if i search for it

**Comment by nephix** [comment]
@andrevmatos can you create a follow up issue or pull request that addresses the feedback, if needed?

File: raiden-ts/src/path/utils.ts (hunk 1 above)
Comment (andrevmatos) [reply raiden-ts/src/path/utils.ts]: Done in #497

File: raiden-ts/src/path/epics.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (andrevmatos) [reply raiden-ts/src/path/epics.ts]: Done in #497
